channeldb/payment_status.go

```diff
@@ -161,7 +184,131 @@ func (ps PaymentStatus) Registrable(settled, paymentFailed bool) error {
+// that's used by the payment router to decide following actions.
+// Together, we use four variables to determine the payment's status,
+//   - inflight: whether there are any pending HTLCs.
+//   - settled: whether any of the HTLCs has been settled.
+//   - failed: whether any of the HTLCs has been failed.
+//   - reason: whether the payment has been marked as failed.
+//
+// Based on the above variables, we derive the status using the following
+// table,
+//
+//	| inflight | settled | failed | reason |         status       |
+//	|:--------:|:-------:|:------:|:------:|:--------------------:|
+//	|   true   |   true  |  true  |   yes  |      StatusError     |
+//	|   true   |   true  |  true  |   no   |    StatusInFlight    |
+//	|   true   |   true  |  false |   yes  |      StatusError     |
+//	|   true   |   true  |  false |   no   |    StatusInFlight    |
+//	|   true   |   false |  true  |   yes  |    StatusInFlight    |
+//	|   true   |   false |  true  |   no   |    StatusInFlight    |
+//	|   true   |   false |  false |   yes  |    StatusInFlight    |
+//	|   true   |   false |  false |   no   |    StatusInFlight    |
+//	|   false  |   true  |  true  |   yes  |      StatusError     |
+//	|   false  |   true  |  true  |   no   |    StatusSucceeded   |
+//	|   false  |   true  |  false |   yes  |      StatusError     |
+//	|   false  |   true  |  false |   no   |    StatusSucceeded   |
+//	|   false  |   false |  true  |   yes  |      StatusFailed    |
+//	|   false  |   false |  true  |   no   |    StatusInFlight    |
+//	|   false  |   false |  false |   yes  |      StatusFailed    |
+//	|   false  |   false |  false |   no   |    StatusInitiated   |
+//
+// Overall, `settled` and `reason` cannot be both true otherwise initiating an
+// unexpected condition. Also, when `inflight`, `settled`, and `failed` are
+// false, this initiates the payment is newly created and hasn't made any HTLCs
+// yet.
+func decidePaymentStatus(htlcs []HTLCAttempt,
+	reason *FailureReason) PaymentStatus {
+
+	var (
+		inflight      bool
+		htlcSettled   bool
+		htlcFailed    bool
+		paymentFailed bool
+	)
+
+	// If we have a failure reason, the payment is failed.
+	if reason != nil {
+		paymentFailed = true
+	}
+
+	// Go through all HTLCs for this payment, check whether we have any
+	// settled HTLC, and any still in-flight.
+	for _, h := range htlcs {
+		if h.Failure != nil {
+			htlcFailed = true
+			continue
+		}
+
+		if h.Settle != nil {
+			htlcSettled = true
+			continue
+		}
+
+		// If any of the HTLCs are not failed nor settled, we
+		// still have inflight HTLCs.
+		inflight = true
+	}
+
+	// Use the DB state to determine the status of the payment.
+	switch {
+	// If any of the HTLCs is settled yet the payment is marked as failed,
+	// this initiates an error state.
+	//
+	// NOTE: whether we have failed or inflight HTLCs is irrelevant.
+	case htlcSettled && paymentFailed:
```

Comment:
Isn't this a mismatch with the table above?

`|   true   |   true  |  true  |   no   |    StatusInFlight    |`

Reply:
This matches the following,
```
//	| inflight | settled | failed | reason |         status       |
//	|:--------:|:-------:|:------:|:------:|:--------------------:|
//	|   true   |   true  |  true  |   yes  |      StatusError     |
//	|   true   |   true  |  false |   yes  |      StatusError     |
//	|   false  |   true  |  true  |   yes  |      StatusError     |
//	|   false  |   true  |  false |   yes  |      StatusError     |
```

Reply:
Ah I interpreted `failed` as `payment failed`. Maybe make it so that table header and code match in terms of naming?

Reply:
cool updated the headers.